sdk/communication/AzureCommunicationChat/Source/Registrar/RegistrarClient.swift

```diff
@@ -0,0 +1,267 @@
+
+    /// Create the data for the a Registrar POST request body.
+    /// - Parameters:
+    ///   - clientDescription: RegistrarClientDescription which is added to the request body.
+    ///   - transports: RegistrarTransports which are added to the body, transport contains APNS push token as the path.
+    private func createPostData(
+        with clientDescription: RegistrarClientDescription,
+        for registrarTransportSettings: [RegistrarTransportSettings]
+    ) -> Data? {
+        let data = RegistrarRequestBody(
+            registrationId: registrationId,
+            nodeId: RegistrarSettings.nodeId,
+            clientDescription: clientDescription,
+            registrarTransportSettings: [
+                RegistrarSettings.pushNotificationTransport: registrarTransportSettings
+            ]
+        )
+
+        return try? JSONEncoder().encode(data)
+    }
+
+    /// Create a Registrar POST request.
+    /// - Parameters:
+    ///   - clientDescription: RegistrarClientDescription.
+    ///   - transports: RegistrarTranports, a transport contains the APNS token as the path.
+    ///   - completionHandler: Returns the POST request.
+    private func createPostRequest(
+        with clientDescription: RegistrarClientDescription,
+        for registrarTransportSettings: [RegistrarTransportSettings],
+        completionHandler: @escaping (HTTPRequest?, AzureError?) -> Void
+    ) {
+        guard let data = createPostData(with: clientDescription, for: registrarTransportSettings) else {
+            completionHandler(nil, AzureError.client("Failed to serialize POST request body."))
+            return
+        }
+
+        guard let request =
+            try? HTTPRequest(method: HTTPMethod.post, url: url, data: data)
+        else {
+            completionHandler(nil, AzureError.client("Failed to create POST request in registration process."))
+            return
+        }
+
+        setRequestHeader(on: request, completionHandler: completionHandler)
+    }
+
+    /// Create a Registrar DELETE request.
+    private func createDeleteRequest(
+        completionHandler: @escaping (HTTPRequest?, AzureError?) -> Void
+    ) {
+        let url = self.url.appendingPathComponent("/\(registrationId)")
+
+        guard let request =
+            try? HTTPRequest(method: HTTPMethod.delete, url: url)
+        else {
+            completionHandler(nil, AzureError.client("Failed to create DELETE request in registration process."))
+            return
+        }
+
+        setRequestHeader(on: request, completionHandler: completionHandler)
+    }
+
+    /// Sets the header on a Registrar request.
+    /// - Parameters:
+    ///   - request: The  HTTP request.
+    ///   - completionHandler: Returns the request, or an error if the request header failed to be created or added
+    private func setRequestHeader(
+        on request: HTTPRequest,
+        completionHandler: @escaping (HTTPRequest?, AzureError?) -> Void
+    ) {
```

Comment:
This is not the correct way to set a header on a PipelineRequest. This should be done as part of the pipeline using a `PipelineStage` object.